### PR TITLE
Adds REST client support for starting and stopping ILM

### PR DESCRIPTION
**Pull request by colings86**
Not sure if this is the best way to expose this API. I think given the REST api for this API are separate endpoints (`_ilm/start` and `_ilm/stop`) that we should either:
1. Split this API into two on the transport layer as well and have a `StartILMAction` and a `StopILMAction` with their own request and response objects
2. Keep the action as is but hide the `PutOperationRequest` from the user and have the client expose a `PutOperationModeResponse startILM(RequestiOptions)` method and a `PutOperationModeResponse stopILM(RequestOptions)` method.

In either case I'm not sure its a good idea to expose the current `PutOperationModeRequest` class to the user since its not particularly user friendly because:
a. As far as the user is concerned what they want to do is start and stop ILM not put it into an "operation mode"
b. The `OperationMode` enum itself is not particular user friendly since its values are `RUNNING`, `STOPPING` and `STOPPED` which relate to the state ILM is in not what you want to do as a user and also because `PutOperationModeRequest` rejects `STOPPED` because you can't go from `RUNNING` straight to `STOPPED` and this is an implementation detail the user should not need to be concerned with when they use this API

Open to thoughts.

### CODE REVIEW

**Comment by colings86** [comment]
Any thoughts on my suggestions above?

**Comment by talevy** [comment]
I prefer (1.) I re-used the same transport request for simplicity under the hood, I do not think this consolidation should be exposed to the user, and users should see specific objects that reflect what they can or cannot do.

**Review by dakrone — COMMENTED** [review]
> Not sure if this is the best way to expose this API. I think given the REST api for this API are separate endpoints (_ilm/start and _ilm/stop) that we should either:
> 1. Split this API into two on the transport layer as well and have a StartILMAction and a StopILMAction with their own request and response objects
> 2. Keep the action as is but hide the PutOperationRequest from the user and have the client expose a PutOperationModeResponse startILM(RequestiOptions) method and a PutOperationModeResponse stopILM(RequestOptions) method.

I think I'm more in favor of option 1, even though it will require slightly more code.

I think this helps us not have to expose the `PutOperationModeRequest` and the Enum that (I agree) is not particularly friendly to a user. So whatever we can do to avoid exposing that would be great.

**Comment by colings86** [comment]
@dakrone @talevy I've pushed a commit which adds StartILMAction and StopILMAction and all the relevant classes to support it on the REST layer and in the REST client. Could you take a look?

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/action/TransportStartILMAction.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.indexlifecycle.action.StartILMAction;
+import org.elasticsearch.xpack.indexlifecycle.OperationModeUpdateTask;
+
+public class TransportStartILMAction extends TransportMasterNodeAction<StartILMRequest, StartILMResponse> {
+
+    @Inject
+    public TransportStartILMAction(Settings settings, TransportService transportService, ClusterService clusterService,
+                                           ThreadPool threadPool, ActionFilters actionFilters,
```
Comment (dakrone) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/action/TransportStartILMAction.java]: Indentation is off here and the line below (super minor)

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/action/TransportStopILMAction.java [hunk 2]
```diff
@@ -17,21 +17,22 @@
...
+import org.elasticsearch.xpack.core.indexlifecycle.action.StopILMAction;
 import org.elasticsearch.xpack.indexlifecycle.OperationModeUpdateTask;
 
-public class TransportPutOperationModeAction extends TransportMasterNodeAction<Request, Response> {
+public class TransportStopILMAction extends TransportMasterNodeAction<StopILMRequest, StopILMResponse> {
 
     @Inject
-    public TransportPutOperationModeAction(Settings settings, TransportService transportService, ClusterService clusterService,
+    public TransportStopILMAction(Settings settings, TransportService transportService, ClusterService clusterService,
                                            ThreadPool threadPool, ActionFilters actionFilters,
```
Comment (dakrone) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/action/TransportStopILMAction.java]: Same here about indentation now

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StartILMRequest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        return null;
+    }
+
+    @Override
+    public int hashCode() {
+        return 64;
```
Comment (dakrone) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StartILMRequest.java]: How did we arrive at this? (not critiquing, just curious)

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StartILMResponse.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * under the License.
+ */
+
+package org.elasticsearch.protocol.xpack.indexlifecycle;
+
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+public class StartILMResponse extends AcknowledgedResponse implements ToXContentObject {
```
Comment (dakrone) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StartILMResponse.java]: Can you remove this class entirely in favor of just returning AcknowledgedResponse? (similar to #32722)

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StopILMRequest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.protocol.xpack.indexlifecycle;
+
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.support.master.AcknowledgedRequest;
+
+public class StopILMRequest extends AcknowledgedRequest<StopILMRequest> {
```
Comment (dakrone) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StopILMRequest.java]: Can you also remove this one in favor of AcknowledgedRequest?

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StopILMResponse.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ * under the License.
+ */
+
+package org.elasticsearch.protocol.xpack.indexlifecycle;
+
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+public class StopILMResponse extends AcknowledgedResponse implements ToXContentObject {
```
Comment (dakrone) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StopILMResponse.java]: Another to remove

**Review by dakrone — CHANGES_REQUESTED** [review]
I left some comments, could you remove the new Response classes and use AcknowledgedResponse everywhere please

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StopILMRequest.java (hunk 5 above)
Comment (dakrone) [reply x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StopILMRequest.java]: Whoops sorry, ignore this, should only change the Response ones

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StartILMRequest.java (hunk 3 above)
Comment (colings86) [reply x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle/StartILMRequest.java]: random number so no method to pick it really. The problem is that we need to be able to do equality so it has to have some value. Ideally I would have an INSTANCE constant and make the constructor private but the Streamable serialisation relies on the constructor being public

**Comment by colings86** [comment]
@dakrone I pushed a commit that should address your comments

**Review by dakrone — APPROVED** [review]
LGTM, thanks Colin!

**Pull request merged**